lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,576 @@
+			node_id_1,
+			node_id_2,
+		);
+		if let Err(lightning_error) = announcement_result {
+			if let ErrorAction::IgnoreDuplicateGossip = lightning_error.action {
+				// everything is fine, just a duplicate channel announcement
+			} else {
+				return Err(lightning_error.into());
+			}
+		}
+	}
+
+	previous_scid = 0; // updates start at a new scid
+
+	let update_count: u32 = Readable::read(read_cursor)?;
+	if update_count < 1 {
+		return Ok(());
+	}
+	let non_incremental_update_presence_flag: u8 = Readable::read(read_cursor)?;
+	let has_non_incremental_updates = non_incremental_update_presence_flag == 1;
+	// without non-incremental updates, the following fields cannot even be calculated by the server
+	let default_cltv_expiry_delta: u16 = if has_non_incremental_updates {
+		Readable::read(&mut read_cursor)?
+	} else {
+		0
+	};
+	let default_htlc_minimum_msat: u64 = if has_non_incremental_updates {
+		Readable::read(&mut read_cursor)?
+	} else {
+		0
+	};
+	let default_fee_base_msat: u32 = if has_non_incremental_updates {
+		Readable::read(&mut read_cursor)?
+	} else {
+		0
+	};
+	let default_fee_proportional_millionths: u32 = if has_non_incremental_updates {
+		Readable::read(&mut read_cursor)?
+	} else {
+		0
+	};
+	let tentative_default_htlc_maximum_msat: u64 = if has_non_incremental_updates {
+		Readable::read(&mut read_cursor)?
+	} else {
+		0
+	};
+	let default_htlc_maximum_msat = if tentative_default_htlc_maximum_msat == u64::MAX {
+		OptionalField::Absent
+	} else {
+		OptionalField::Present(tentative_default_htlc_maximum_msat)
+	};
+
+	for _current_update_index in 0..update_count {
+		// println!("update index: {}", _current_update_index);
+
+		let scid_delta: BigSize = Readable::read(read_cursor)?;
+		let short_channel_id = previous_scid + scid_delta.0;
+		previous_scid = short_channel_id;
+
+		// let short_channel_id: u64 = Readable::read(read_cursor)?;
+		let channel_flags: u8 = Readable::read(read_cursor)?;
+
+		// flags are always sent in full, and hence always need updating
+		let standard_channel_flags = channel_flags & 0b_0000_0011;
+
+		let mut synthetic_update = if channel_flags & 0b_1000_0000 == 0 {
+			// full update
+			// println!("channel flags: {:b}, scid: {}, full update", channel_flags, short_channel_id);
+			UnsignedChannelUpdate {
+				chain_hash,
+				short_channel_id,
+				timestamp: emulated_last_seen_timestamp,
+				flags: standard_channel_flags,
+				cltv_expiry_delta: default_cltv_expiry_delta,
+				htlc_minimum_msat: default_htlc_minimum_msat,
+				htlc_maximum_msat: default_htlc_maximum_msat.clone(),
+				fee_base_msat: default_fee_base_msat,
+				fee_proportional_millionths: default_fee_proportional_millionths,
+				excess_data: vec![],
+			}
+		} else {
+			// incremental update
+			// println!("channel flags: {:b}, incremental update", channel_flags);
+
+			let read_only_network_graph = network_graph.read_only();
+			let channel = read_only_network_graph
+				.channels()
+				.get(&short_channel_id)
```

Comment:
If we have performance concerns with the `BTreeMap`, one possibility would be to traverse the (sorted) channel updates in parallel with the tree rather than performing look-ups for each update. That would be quite an involved change though.